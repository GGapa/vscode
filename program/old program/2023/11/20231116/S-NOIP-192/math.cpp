#include <bits/stdc++.h>
using namespace std;
#define int __int128
typedef pair<int, int> pii;

int n, sum;
pii B, A, C;

inline int read()
{
    int x=0,f=1;
    char ch=getchar();
    while(ch<'0'||ch>'9')
    {
        if(ch=='-')
            f=-1;
        ch=getchar();
    }
    while(ch>='0' && ch<='9')
        x=x*10+ch-'0',ch=getchar();
    return x*f;
}

void write(int x)
{
    if(x<0)
        putchar('-'),x=-x;
    if(x>9)
        write(x/10);
    putchar(x%10+'0');
    return;
}

int gcd(int a,int b) {
    return b > 0 ? gcd(b, a % b) : a;
}

void easy(pii& x) {
    bool flag = (x.first * x.second < 0);
    x.first = x.first < 0 ? -x.first : x.first;
    x.second = x.second < 0 ? -x.second : x.second;
    int gd = gcd(x.first, x.second);
    x.first /= gd; x.second /= gd;
    if(flag) x.first *= -1;
}

pii add(pii x, pii y) {
    x.first = x.first * y.second + y.first * x.second;
    x.second = x.second * y.second;
    easy(x);
    return x;
}
pii sub(pii x, pii y) {
    x.first = x.first * y.second - y.first * x.second, x.second *= y.second;
    easy(x);
    return x;
}


pii pow2(pii x) {
    x.first *= x.first; x.second *= x.second;
    easy(x);
    return x;
}

pii mul(pii x, int y) {
    if(y == 0) return make_pair(0, 1);
    if(x.second % y == 0) x.second /= y, y = 1;
    else {
        x.first *= y;
    }
    easy(x);
    return x;
}

pii dvi(pii x, int y) {
    x.second *= y;
    easy(x);
    return x;
}

void out() {
    easy(A); easy(B);
    write(sum); 
    putchar(' ');
    if((A.first) % (A.second) == 0) write(A.first / A.second);
    else write(A.first), putchar('/'), write(A.second);
    putchar(' ');
    if(B.first % B.second == 0) write(B.first / B.second);
    else write(B.first), putchar('/'), write(B.second);
    putchar('\n');
}   

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0); 
// #ifdef ONLINE_JUDGE
    freopen("math.in", "r", stdin);
    freopen("math.out", "w", stdout);
// #endif
    n = read(); 
    A.second = B.second = 1;
    for(int i = 1, x; i <= n; i++) {
        x = read();
        sum += x;
        A.first = sum; A.second = i;
        B = add(B, pow2(make_pair(x, 1)));
        C = B;
        B = sub(dvi(B, i), pow2(A));
        out();
        B = C;
    }
    return 0;
}