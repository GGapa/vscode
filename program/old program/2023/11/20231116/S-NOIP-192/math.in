500
724 940 339 282 563 304 329 630 401 616 780 123 444 134 674 50 425 324 127 870 599 303 220 237 523 505 715 732 907 565 419 818 387 341 585 747 926 6 172 254 183 188 89 191 167 552 856 163 349 492 609 203 126 520 539 534 586 649 23 350 433 670 653 107 584 373 571 622 327 536 161 321 822 453 1000 121 86 753 519 986 352 381 476 457 527 755 284 274 45 845 71 316 299 333 174 198 219 830 317 584 557 530 414 297 264 533 842 225 300 237 66 30 556 362 233 715 123 926 783 81 939 924 142 951 249 754 621 982 483 536 23 640 877 417 363 89 247 216 895 372 545 489 45 668 388 433 892 587 669 811 693 171 646 870 584 419 694 164 304 512 282 485 697 781 322 780 226 440 560 500 230 198 401 557 735 939 44 325 581 681 80 700 985 369 800 774 169 496 171 761 311 57 153 126 414 475 649 990 666 529 313 526 953 730 264 221 924 664 394 1 275 491 649 205 574 867 67 696 796 39 3 528 183 2 13 518 20 858 329 878 531 276 167 160 484 916 752 251 667 542 637 349 908 869 4 681 522 779 503 842 328 503 880 45 389 608 838 34 515 68 191 300 928 80 399 552 517 189 488 134 715 253 50 32 923 169 527 857 929 724 48 245 10 801 562 268 515 668 542 511 830 418 635 249 210 731 280 365 496 681 939 228 660 209 170 858 534 794 196 468 854 750 612 613 823 251 921 701 757 165 468 820 944 97 687 824 305 513 71 247 906 333 112 301 30 634 975 529 642 442 29 411 222 122 536 728 457 810 532 980 311 545 289 204 83 801 997 476 974 405 613 493 927 533 796 315 348 583 861 462 518 85 313 468 997 827 740 62 653 261 329 642 10 855 906 844 418 321 521 856 749 350 90 130 740 278 893 775 334 860 433 46 767 67 434 836 693 646 491 363 445 681 694 134 571 650 888 692 445 133 225 196 48 482 351 311 132 675 829 44 241 594 772 622 587 774 457 703 193 585 924 882 757 71 804 823 544 12 597 915 418 322 40 217 456 862 23 625 739 907 385 426 233 275 910 340 556 337 580 18 386 957 6 379 604 669 399 756 620 436 410 344 331 751 778 649 542 411 378 674 947 160 981 655 944 968 760 317 536 248 
