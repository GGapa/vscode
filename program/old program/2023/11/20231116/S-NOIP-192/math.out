724 724 0
1664 832 11664
2003 2003/3 556082/9
2285 2285/4 1187659/16
2848 2848/5 1484846/25
3152 1576/3 533633/9
3481 3481/7 2721568/49
4111 4111/8 3233655/64
4512 1504/3 415532/9
5128 2564/5 1068414/25
5908 5908/11 5414980/121
6031 6031/12 7793435/144
6475 6475/13 8484072/169
6609 6609/14 10859869/196
7283 7283/15 12206426/225
7333 7333/16 15865527/256
7758 7758/17 16874878/289
8082 449 504592/9
8209 8209/19 21040808/361
9079 9079/20 25792379/400
9678 3226/7 9167596/147
9981 9981/22 29335741/484
10201 10201/23 31870542/529
10438 5219/12 8559299/144
10961 10961/25 35849954/625
11466 441 719129/13
12181 12181/27 40784942/729
12913 12913/28 44425195/784
13820 13820/29 51576998/841
14385 959/2 3572159/60
14804 14804/31 55478272/961
15622 7811/16 15215255/256
16009 16009/33 63090572/1089
16350 8175/17 16421962/289
16935 3387/7 13597684/245
17682 2947/6 669949/12
18608 18608/37 81171220/1369
18614 9307/19 23125347/361
18786 6262/13 10974948/169
19040 476 645807/10
19223 19223/41 109346308/1681
19411 2773/6 16463903/252
19500 19500/43 123839946/1849
19691 19691/44 129682651/1936
19858 19858/45 136092476/2025
20410 10205/23 34917079/529
21266 21266/47 150524348/2209
21429 7143/16 52555069/768
21778 21778/49 161405614/2401
22270 2227/5 1648104/25
22879 22879/51 169444856/2601
23082 11541/26 43960947/676
23208 23208/53 184480032/2809
23728 11864/27 47079536/729
24267 24267/55 192341126/3025
24801 3543/8 28044529/448
25387 25387/57 200964416/3249
26036 13018/29 51713313/841
26059 26059/59 220940184/3481
26409 8803/20 25020091/400
26842 26842/61 228936900/3721
27512 13756/31 58978986/961
28165 28165/63 242435954/3969
28272 1767/4 1981013/32
28856 28856/65 258826734/4225
29229 9743/22 29237309/484
29800 29800/67 268206336/4489
30422 15211/34 68578443/1156
30749 30749/69 279333248/4761
31285 6257/14 56788993/980
31446 31446/71 293724182/5041
31767 10589/24 33212911/576
32589 32589/73 313507978/5329
33042 16521/37 79451441/1369
34042 34042/75 344770136/5625
34163 34163/76 357678419/5776
34249 34249/77 372427492/5929
35002 17501/39 96144650/1521
35521 35521/79 389894100/6241
36507 36507/80 417556951/6400
36859 36859/81 423647318/6561
37240 18620/41 107330420/1681
37716 37716/83 434596472/6889
38173 38173/84 439833131/7056
38700 7740/17 89102298/1445
39455 39455/86 458387809/7396
39739 39739/87 466345004/7569
40013 40013/88 474611519/7744
40058 40058/89 494775500/7921
40903 40903/90 514214681/8100
40974 40974/91 533163146/8281
41290 20645/46 135165629/2116
41589 13863/31 182867962/2883
41922 20961/47 138928897/2209
42096 42096/95 568580864/9025
42294 7049/16 48356141/768
42513 42513/97 591030844/9409
43343 43343/98 612008193/9604
43660 43660/99 619791182/9801
44244 11061/25 78509483/1250
44801 44801/101 635669022/10201
45331 45331/102 642717185/10404
45745 45745/103 649112732/10609
46042 23021/52 164411087/2704
46306 46306/105 667289384/11025
46839 46839/106 674533057/11236
47681 47681/107 697866978/11449
47906 23953/54 177399245/2916
48206 48206/109 718393582/11881
48443 48443/110 729576561/12100
48509 48509/111 751627616/12321
48539 48539/112 776787703/12544
49095 49095/113 785407194/12769
49457 2603/6 41734271/684
49690 9938/23 160900988/2645
50405 50405/116 820705211/13456
50528 50528/117 839037866/13689
51454 25727/59 218694288/3481
52237 52237/119 896394616/14161
52318 26159/60 229793999/3600
53257 53257/121 957198888/14641
54181 54181/122 993438145/14884
54323 54323/123 1012715822/15129
55274 27637/62 263215009/3844
55523 55523/125 1066151346/15625
56277 6253/14 40247363/588
56898 56898/127 1099133688/16129
57880 7235/16 17875015/256
58363 58363/129 1153059992/16641
58899 58899/130 1162899449/16900
58922 58922/131 1195889570/17161
59562 9927/22 100813187/1452
60439 3181/7 65413282/931
60856 30428/67 313095858/4489
61219 61219/135 1262842874/18225
61308 15327/34 40316589/578
61555 61555/137 1305265462/18769
61771 61771/138 1322250125/19044
62666 62666/139 1359452692/19321
63038 31519/70 342524199/4900
63583 63583/141 1381139492/19881
64072 32036/71 347784755/5041
64117 64117/143 1424366852/20449
64785 21595/48 53378717/768
65173 65173/145 1451785546/21025
65606 32803/73 365459297/5329
66498 22166/49 500152024/7203
67085 67085/148 1513327771/21904
67754 67754/149 1530440346/22201
68565 4571/10 1386333/20
69258 69258/151 1578369444/22801
69429 69429/152 1601317407/23104
70075 70075/153 1617295220/23409
70945 10135/22 236262255/3388
71529 71529/155 1666916914/24025
71948 17987/39 209743861/3042
72642 72642/157 1697161162/24649
72806 36403/79 430494444/6241
73110 24370/53 192973430/2809
73622 36811/80 437029239/6400
73904 73904/161 1764119962/25921
74389 74389/162 1775185805/26244
75086 75086/163 1795305320/26569
75867 75867/164 1823047163/26896
76189 76189/165 1837405484/27225
76969 76969/166 1865252821/27556
77195 77195/167 1885866026/27889
77635 77635/168 1897241279/28224
78195 6015/13 11302628/169
78695 15739/34 384336397/5780
78925 78925/171 1942208138/29241
79123 79123/172 1965443475/29584
79524 79524/173 1977469558/29929
80081 80081/174 1990538657/30276
80816 80816/175 2015114744/30625
81755 81755/176 2066479703/30976
81799 81799/177 2109343952/31329
82124 41062/89 531147529/7921
82705 82705/179 2139073400/32041
83386 41693/90 539901371/8100
83466 83466/181 2198042560/32761
84166 42083/91 555128349/8281
85151 85151/183 2282410562/33489
85520 10690/23 143536253/2116
86320 17264/37 465947534/6845
87094 43547/93 589953248/8649
87263 7933/17 217196024/3179
87759 87759/188 2402093619/35344
87930 9770/21 30016304/441
88691 88691/190 2460717549/36100
89002 89002/191 2478280374/36481
89059 89059/192 2523203063/36864
89212 89212/193 2554897144/37249
89338 44669/97 647488701/9409
89752 6904/15 200286512/2925
90227 90227/196 2617119419/38416
90876 90876/197 2637448082/38809
91866 15311/33 25054652/363
92532 92532/199 2727650206/39601
93061 93061/200 2742172479/40000
93374 93374/201 2760522704/40401
93900 46950/101 693753928/10201
94853 94853/203 2836887786/41209
95583 31861/68 318319619/4624
95847 95847/205 2887455096/42025
96068 48034/103 728500297/10609
96992 96992/207 2971292282/42849
97656 939/2 7196041/104
98050 98050/209 3009130790/43681
98051 98051/210 3069331709/44100
98326 466 14652520/211
98817 98817/212 3106466115/44944
99466 99466/213 3128209778/45369
99671 99671/214 3157514761/45796
100245 20049/43 634955412/9245
101112 4213/9 11194693/162
101179 101179/217 3273750108/47089
101875 101875/218 3300289609/47524
102671 102671/219 3338979710/47961
102710 10271/22 33946849/484
102713 7901/17 265957890/3757
103241 103241/222 3473980865/49284
103424 103424/223 3507289960/49729
103426 51713/112 892642847/12544
103439 103439/225 3631614404/50625
103957 103957/226 3648518905/51076
103977 103977/227 3708413754/51529
104835 34945/76 417895731/5776
105164 105164/229 3781458390/52441
106042 53021/115 959532594/13225
106573 106573/231 3855943220/53361
106849 106849/232 3880571959/53824
107016 107016/233 3917291196/54289
107176 53588/117 988743830/13689
107660 21532/47 158881396/2209
108576 27144/59 252387765/3481
109328 109328/237 4075428236/56169
109579 109579/238 4103105677/56644
110246 110246/239 4130502734/57121
110788 27697/60 259333901/3600
111425 111425/241 4174013578/58081
111774 55887/121 1048607313/14641
112682 112682/243 4259926040/59049
113551 113551/244 4317371395/59536
113555 22711/49 175480184/2401
114236 57118/123 1104125465/15129
114758 114758/247 4435272026/61009
115537 3727/8 144440089/1984
116040 38680/83 1498679860/20667
116882 58441/125 1137323519/15625
117210 117210/251 4572358268/63001
117713 117713/252 4590900635/63504
118593 118593/253 4652077932/64009
118638 59319/127 1178973666/16129
119027 119027/255 4736009636/65025
119635 119635/256 4759668247/65536
120473 120473/257 4813435238/66049
120507 40169/86 1626914399/22188
121022 121022/259 4900253224/67081
121090 12109/26 248023063/3380
121281 40427/87 555462664/7569
121581 121581/262 5025395893/68644
122509 122509/263 5100972282/69169
122589 40863/88 573279527/7744
122988 122988/265 5180186886/70225
123540 61770/133 1300445499/17689
124057 124057/267 5222072516/71289
124246 62123/134 1315478943/17956
124734 124734/269 5281709284/72361
124868 62434/135 1332645974/18225
125583 125583/271 5367544522/73441
125836 31459/68 337459269/4624
125886 41962/91 607279006/8281
125918 62959/137 1383950688/18769
126841 11531/25 510441504/6875
127010 63505/138 1414690091/19044
127537 127537/277 5680495392/76729
128394 64197/139 1436141860/19321
129323 129323/279 5825899328/77841
130047 130047/280 5865710311/78400
130095 130095/281 5935220678/78961
130340 65170/141 1492423304/19881
130350 130350/283 6048526704/80089
131151 131151/284 6102691275/80656
131713 131713/285 6127031036/81225
131981 131981/286 6159272341/81796
132496 18928/41 883089674/11767
133164 3699/8 43162403/576
133706 133706/289 6238793194/83521
134217 134217/290 6261056301/84100
135047 135047/291 6321744872/84681
135465 135465/292 6344086955/85264
136100 136100/293 6374359520/85849
136349 136349/294 6409722653/86436
136559 136559/295 6450458114/87025
137290 68645/148 1623381523/21904
137570 137570/297 6525465194/88209
137935 137935/298 6550300421/88804
138431 138431/299 6572608396/89401
139112 34778/75 413050166/5625
140051 3257/7 155781468/2107
140279 929/2 44621227/604
140939 140939/303 6771658406/91809
141148 35287/76 425867941/5776
141318 141318/305 6862631826/93025
142176 23696/51 192573293/2601
142710 142710/307 6956766870/94249
143504 35876/77 876585869/11858
143700 47900/103 2352631886/31827
144168 72084/155 1770184859/24025
145022 145022/311 7150476038/96721
145772 36443/78 224953045/3042
146384 146384/313 7228109642/97969
146997 146997/314 7257812501/98596
147820 29564/63 1464093292/19845
148071 148071/316 7358713179/99856
148992 148992/317 7446680598/100489
149693 149693/318 7487086217/101124
150450 150450/319 7536690284/101761
150615 30123/64 1518061859/20480
151083 50361/107 2538010432/34347
151903 151903/322 7676924561/103684
152847 8991/19 457210522/6137
152944 38236/81 980294773/13122
153631 153631/325 7881533014/105625
154455 154455/326 7945890225/106276
154760 154760/327 7979551700/106929
155273 155273/328 8004470103/107584
155344 22192/47 164938446/2209
155591 155591/330 8123229989/108900
156497 156497/331 8210150096/109561
156830 78415/166 2060355827/27556
156942 17438/37 923262760/12321
157243 157243/334 8343975277/111556
157273 157273/335 8433851286/112225
157907 157907/336 8468095271/112896
158882 158882/337 8578999466/113569
159411 159411/338 8605572197/114244
160053 53351/113 2880281072/38307
160495 32099/68 1733328047/23120
160524 160524/341 8758868304/116281
160935 53645/114 976196813/12996
161157 161157/343 8832591906/117649
161279 161279/344 8899844695/118336
161815 32363/69 357090728/4761
162543 162543/346 8976282105/119716
163000 163000/347 9002281606/120409
163810 81905/174 2267099801/30276
164342 164342/349 9095764714/121801
165322 82661/175 2303071004/30625
165633 4247/9 6080024/81
166178 83089/176 2318984863/30976
166467 166467/353 9314092128/124609
166671 55557/118 1040639081/13924
166754 166754/355 9445452354/126025
167555 167555/356 9511016995/126736
168552 56184/119 3212119108/42483
169028 84514/179 2415838814/32041
170002 170002/359 9780513100/128881
170407 170407/360 9809443511/129600
171020 171020/361 9843712450/130321
171513 171513/362 9871114269/131044
172440 57480/121 1108081796/14641
172973 172973/364 10001428659/132496
173769 173769/365 10066365184/133225
174084 29014/61 841951231/11163
174432 174432/367 10136982530/134689
175015 175015/368 10168861343/135424
175876 175876/369 10251158792/136161
176338 88169/185 2569754669/34225
176856 176856/371 10307434304/137641
176941 176941/372 10392139499/138384
177254 177254/373 10429916366/139129
177722 88861/187 2614474502/34969
178719 59573/125 3529231388/46875
179546 89773/188 2665493695/35344
180286 180286/377 10716236696/142129
180348 4294/9 42896708/567
181001 181001/379 10850262370/143641
181262 90631/190 2724166989/36100
181591 181591/381 10933667510/145161
182233 182233/382 10972785761/145924
182243 182243/383 11084838076/146689
183098 91549/192 2792210999/36864
184004 184004/385 11268661344/148225
184848 92424/193 2837380672/37249
185266 185266/387 11380356296/149769
185587 185587/388 11419390187/150544
186108 186108/389 11449528478/151321
186964 93482/195 2883604541/38025
187713 187713/391 11592341496/152881
188063 188063/392 11628605911/153664
188153 188153/393 11717818286/154449
188283 188283/394 11795436821/155236
189023 189023/395 11852445896/156025
189301 189301/396 11898337403/156816
190194 190194/397 11996574020/157609
190969 190969/398 12061557293/158404
191303 27329/57 1728617966/22743
192163 192163/400 12188433031/160000
192596 192596/401 12219803102/160801
192642 32107/67 1027158992/13467
193409 6239/13 399673048/5239
193476 48369/101 780582642/10201
193910 38782/81 2504210188/32805
194746 97373/203 3150911167/41209
195439 195439/407 12653165070/165649
196085 196085/408 12695443007/166464
196576 196576/409 12726603414/167281
196939 196939/410 12763378669/168100
197384 197384/411 12795020888/168921
198065 198065/412 12842715355/169744
198759 198759/413 12892624632/170569
198893 198893/414 12973644149/171396
199464 199464/415 13008378374/172225
200114 100057/208 3262906943/43264
201002 201002/417 13151897210/173889
201694 100847/209 3300455718/43681
202139 202139/419 13233994668/175561
202272 2408/5 13211052/175
202497 202497/421 13376101236/177241
202693 202693/422 13442066893/178084
202741 202741/423 13552790414/178929
203223 203223/424 13584833207/179776
203574 203574/425 13623852274/180625
203885 203885/426 13667903261/181476
204017 204017/427 13751164556/182329
204692 51173/107 862498451/11449
205521 68507/143 4628290796/61347
205565 41113/86 2799688337/36980
205806 205806/431 14055160722/185761
206400 4300/9 48936181/648
207172 207172/433 14163641070/187489
207794 103897/217 3551318328/47089
208381 208381/435 14243086424/189225
209155 209155/436 14313675587/190096
209612 209612/437 14346730048/190969
210315 70105/146 4800452575/63948
210508 210508/439 14470358222/192721
211093 211093/440 14508204911/193600
212017 212017/441 14628012986/194481
212899 212899/442 14732179757/195364
213656 213656/443 14799016528/196249
213727 213727/444 14907361787/197136
214531 214531/445 14987153764/198025
215354 107677/223 3768137462/49729
215898 71966/149 5036004062/66603
215910 107955/224 3810242743/50176
216507 216507/449 15280921778/201601
217422 12079/25 427751681/5625
217840 217840/451 15435191272/203401
218162 109081/226 3870277067/51076
218202 72734/151 5201308888/68403
218419 218419/454 15670108069/206116
218875 43775/91 3140981946/41405
219737 219737/456 15805459031/207936
219760 219760/457 15936140178/208849
220385 220385/458 15980504157/209764
221124 73708/153 1782870836/23409
222031 222031/460 16163799579/211600
222416 222416/461 16203326958/212521
222842 111421/231 4059986234/53361
223075 223075/463 16303819532/214369
223350 111675/232 4089708599/53824
224260 44852/93 3295868632/43245
224600 112300/233 4131048956/54289
225156 225156/467 16562209064/218089
225493 225493/468 16607510843/219024
226073 226073/469 16647507932/219961
226091 226091/470 16783991429/220900
226477 226477/471 16823947790/221841
227434 113717/236 4241613795/55696
227440 227440/473 17109278302/223729
227819 227819/474 17150356349/224676
228423 228423/475 17193752796/225625
229092 3369/7 63407199/833
229491 25499/53 640230140/8427
230247 230247/478 17358496653/228484
230867 230867/479 17403955702/229441
231303 77101/160 5813767397/76800
231713 231713/481 17480118356/231361
232057 232057/482 17525584153/232324
232388 232388/483 17572853882/233289
233139 233139/484 17644412211/234256
233917 233917/485 17723362006/235225
234566 117283/243 4443345533/59049
235108 235108/487 17811665000/237169
235519 235519/488 17850747623/238144
235897 235897/489 17892668444/239121
236571 236571/490 17947208929/240100
237518 237518/491 18089422818/241081
237678 39613/82 504936845/6724
238659 238659/493 18336649158/243049
239314 119657/247 4597060729/61009
240258 80086/165 2058866224/27225
241226 120613/248 4670632823/61504
241986 241986/497 18757342280/247009
242303 242303/498 18809428697/248004
242839 242839/499 18848416286/249001
243087 243087/500 18914608931/250000
