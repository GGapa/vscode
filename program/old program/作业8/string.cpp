/*
#include <iostream>
#include <string>
#define chu 1000000007;
using namespace std;

string T;
int ans = 0, n, m;
string S, R;
void digui(string a, string b, int f)
{
	if (b == T)
	{
		ans++;
		ans %= chu;
		a = S; b = "";
		return;
	}
	if (f == 1)
	{
		while (a != "")
		{
			if (a[0] == '-')
			{
				b.erase(b.length()-1);
				a.erase(0, 1);
				digui(a, b, 0);
			}
			else
			{
				b += a[0];
				a.erase(0, 1);
			}
		}
	}
	else
	{
		while (a != "")
		{
			if (a[0] == '-')
			{
				b[0] += a[0];
				b.erase(0, 1);
				a.erase(0, 1);
				digui(a, b, 0);
			}
			else
			{
				b += a[0];
				a.erase(0, 1);
			}
		}
	}
}
int main()
{
	int t;
	cin >> t;
	for (int i = 0; i < t; i++)
	{
		cin >> n >> m>>S>>T;
		digui(S, R, 1);
		cout << ans << endl;
		ans = 0;
	}
	return 0;
}
*/