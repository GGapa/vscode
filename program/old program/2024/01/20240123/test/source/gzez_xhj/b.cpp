#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <algorithm>
#include <vector>
#include <set>
typedef long long lld;
using namespace std;
#define fir first
#define sec second
const int mod = 1e9 + 7;
const int mod1 = 1e9 + 7;
const int mod2 = 998244353;
const int base = 1453;
const int N = 810;
inline int qpow(int a, int b, int p) {
	int ret = 1;
	for (; b; b >>= 1, a = 1ll * a * a % p)
		if (b & 1) ret = 1ll * ret * a % p;
	return ret;
}
int n, m;
struct tree {
	vector<int> G[N], ch[N];
	int cnt[N];
	inline void addedge(int u, int v) {
		G[u].push_back(v);
		G[v].push_back(u);
	}
	int fa[N], dep[N], siz[N];
	inline void dfs0(int u, int f) {
		dep[u] = dep[fa[u] = f] + (siz[u] = 1);
		if (f) ch[f].push_back(u);
		for (int v : G[u]) if (v != f) dfs0(v, u), siz[u] += siz[v];
	}
	pair<int, int> hash[N];
	inline pair<int, int> hash_tree(int u) {
		pair<int, int> ret = make_pair(siz[u], siz[u]), res;
		for (int v : ch[u]) {
			res = hash_tree(v);
			ret.fir = (1ll * ret.fir * base % mod1 + res.fir) % mod1;
			ret.sec = (1ll * ret.sec * base % mod2 + res.sec) % mod2;
		}
		return hash[u] = ret;
	}
} T1, T2, t1, t2;
inline void build(const tree &T, tree &t, int u, int tp) {
	++ t.cnt[tp];
	if (T.ch[u].size()) {
		if (T.ch[u].size() == 1) build(T, t, T.ch[u][0], tp);
		else for (int v : T.ch[u]) build(T, t, v, v), t.addedge(tp, v);
	}
}
int to[N];
inline bool calcto(int u, int v) {
	if (t1.ch[u].size() != t2.ch[v].size()) return 0;
	to[u] = v;
	for (int i = 0; i < t1.ch[u].size(); ++ i)
		if (!calcto(t1.ch[u][i], t2.ch[v][i])) return 0;
	return 1;
}
set<int> s; vector<int> v;
int f[N][N]; int vis[N][N];
inline int calc(int i, int j, int t) {
	if (!i) return !j;
	if (i > j) return 0;
	if (vis[i][j] == t) return f[i][j];
	vis[i][j] = t;
	int ret = 0;
	for (int k : v)
		if (k > j) break;
		else ret = (ret + calc(i - 1, j - k, t)) % mod;
	return f[i][j] = ret;
}
int main() {
//	freopen("ex_b3.in", "r", stdin);
	freopen("b.in", "r", stdin);
	freopen("b.out", "w", stdout);
	scanf("%d%d", &n, &m);
	for (int i = 2, p; i <= n; ++ i) scanf("%d", &p), T1.addedge(i, p);
	for (int i = 2, p; i <= m; ++ i) scanf("%d", &p), T2.addedge(i, p);
	T1.dfs0(1, 0), T2.dfs0(1, 0);
	build(T1, t1, 1, 1), t1.dfs0(1, 0);
	build(T2, t2, 1, 1), t2.dfs0(1, 0);
	int q;
	scanf("%d", &q);
	if (!calcto(1, 1)) {
		for (; q --; ) puts("0");
		return 0;
	}
	int mx1 = 0, mx2 = 0;
	for (int i = 1; i <= n; ++ i) mx1 = max(mx1, t1.cnt[i]);
	for (int i = 1; i <= n; ++ i) mx2 = max(mx2, t2.cnt[to[i]]);
	for (int op, x, cs = 1; cs <= q; ++ cs) {
		scanf("%d%d", &op, &x);
		if (op == 1) s.insert(x); else s.erase(x);
		v.clear(); for (int i : s) v.push_back(i);
		int ans = 1;
		for (int i = 1; i <= n; ++ i) {
			ans = 1ll * ans * calc(t1.cnt[i], t2.cnt[to[i]], cs) % mod;
			if (!ans) break;
		}
		printf("%d\n", ans);
	}
	fclose(stdin);
	fclose(stdout);
	return 0;
}
