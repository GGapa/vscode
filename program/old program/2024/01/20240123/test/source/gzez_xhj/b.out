0
0
248160799
946824203
934862950
938867606
515738879
922231092
967179605
737891055
499002000
193179029
346241718
184945186
324959253
753054279
599221605
691613092
797397672
578225987
788584264
75597343
858865259
382700766
699448967
711640606
614576898
494160192
719924434
721902579
891387353
861123254
672647366
625219060
981458993
465158462
774648850
984732160
198773238
423548703
957398903
359285180
448490182
321319094
279275317
713788983
789440886
17154030
196231165
112973491
417962775
907978703
660873523
148051949
452428263
469280396
